pysgrid/sgrid.py

```diff
@@ -252,13 +275,491 @@ def _save_common_components(self, nc_file):
+
+        if self._trees[grid] is None:
+            self.build_celltree(grid)
+        tree = self._trees[grid][0]
+        indices = tree.locate(points)
+        lon, lat = self._get_grid_vars(grid)
+        x = indices % (lat.shape[1] - 1)
+        y = indices // (lat.shape[1] - 1)
+        ind = np.column_stack((y, x))
+        ind[ind[:, 0] == -1] = [-1, -1]
+        if just_one:
+            res = ind[0]
+            return res
+        else:
+            res = np.ma.masked_less(ind, 0)
+            if _memo:
+                self._add_memo(points, res, grid,
+                               self._ind_memo_dict, _copy, _hash)
+            return res
+
+    def get_variable_by_index(self, var, index):
+        """
+        Function to get the node values of a given face index.
+        Emulates the 'self.grid.nodes[self.grid.nodes.faces[index]]'
+        paradigm of unstructured grids.
+        """
+
+        arr = var[:]
+        x = index[:, 0]
+        y = index[:, 1]
+        return np.ma.column_stack((arr[x, y], arr[x + 1, y],
+                                   arr[x + 1, y + 1], arr[x, y + 1]))
+
+    def build_celltree(self, grid='node'):
+        """
+        Tries to build the celltree for grid defined by the
+        node coordinates of the specified grid.
+
+        :param grid: which grid to biuld the celltree for. Options are:
+                     'node', 'edge1', 'edge2', 'center'
+        """
+
+        if not hasattr(self, '_ind_memo_dict'):
+            self._ind_memo_dict = {'node': None,
+                                   'edge1': None,
+                                   'edge2': None,
+                                   'center': None}
+        if not hasattr(self, '_trees'):
+            self._trees = {'node': None,
+                           'edge1': None,
+                           'edge2': None,
+                           'center': None}
+        try:
+            from cell_tree2d import CellTree
+        except ImportError:
+            raise ImportError("the cell_tree2d package must be installed to "
+                              "use the celltree search:\n"
+                              "https://github.com/NOAA-ORR-ERD/cell_tree2d/")
+
+        lon, lat = self._get_grid_vars(grid)
+        if lon is None or lat is None:
+            raise ValueError(
+                "{0}_lon and {0}_lat must be defined in order to create and "
+                "use CellTree for this grid".format(grid))
+        lin_nodes = np.ascontiguousarray(np.column_stack((lon[:].reshape(-1), lat[:].reshape(-1)))).astype(np.float64)  # noqa
+        y_size = lon.shape[0]
+        x_size = lon.shape[1]
+        lin_faces = np.array([np.array([[x, x + 1, x + x_size + 1, x + x_size]
+                                        for x in range(0, x_size - 1, 1)]) +
+                              y * x_size for y in range(0, y_size - 1)])
+        lin_faces = np.ascontiguousarray(lin_faces.reshape(-1, 4).astype(np.int32))  # noqa
```

Comment:
Some as https://github.com/sgrid/pysgrid/pull/93/files#r71764247 but the `int32`.